### PR TITLE
etcdctl: member add/update accept peerURLs prefix

### PR DESCRIPTION
**Pull request by xiang90**
Partially fix https://github.com/coreos/etcd/issues/3476.

We can not force people to add `peerURLs=` for backward compatibility issue.
I will update doc and make this more clear when this pr get merged.

/cc @robszumski

### CODE REVIEW

**Comment by gyuho** [comment]
Looks good to me. Yeah we don't have any man page on this for now:

```
./etcdctl member add -h
NAME:
   add - add a new member to the etcd cluster

USAGE:
   command add [arguments...]
```

**Comment by xiang90** [comment]
@gyuho Right... We might start to fill in the man page. But not a priority for now.

File: etcdctl/command/member_commands.go [hunk 1]
```diff
@@ -186,3 +197,15 @@ func actionMemberUpdate(c *cli.Context) {
 
 	fmt.Printf("Updated member with ID %s in cluster\n", mid)
 }
+
+func getPeerURLsStringFromArg(arg string) (string, error) {
```
Comment (jonboulle) [inline etcdctl/command/member_commands.go]: Comment explaining why this is necessary would be handy.

File: etcdctl/command/member_commands.go [hunk 2]
```diff
@@ -186,3 +197,15 @@ func actionMemberUpdate(c *cli.Context) {
...
+	segs := strings.Split(arg, "=")
```
Comment (jonboulle) [inline etcdctl/command/member_commands.go]: `SplitN(arg, "=", 1)` ?

File: etcdctl/command/member_commands.go [hunk 3]
```diff
@@ -82,8 +82,14 @@ func actionMemberAdd(c *cli.Context) {
 	}
 
 	mAPI := mustNewMembersAPI(c)
+	var url string
```
Comment (jonboulle) [inline etcdctl/command/member_commands.go]: No point in declaring if you're doing so on line 87 too

**Comment by jonboulle** [comment]
Devil's advocate, if this is a command that would only ever be run manually (?), do we need to worry about backwards compatibility? Maybe this is an opportunity to just clean up the UX?

**Comment by eyakubovich** [comment]
If you start having named args (key=val), maybe just use the usual `--key=val` format. Not only will it save manual parsing, it'll feel more natural when you start adding other keys and making them optional.

**Comment by xiang90** [comment]
@eyakubovich I would like to do that but it breaks the previous experience pretty heavily (we do not use flag for actual configuration).

**Comment by xiang90** [comment]
> Devil's advocate, if this is a command that would only ever be run manually (?), do we need to worry about backwards compatibility? Maybe this is an opportunity to just clean up the UX?

Some people run etcdctl programmatically. 

> Maybe this is an opportunity to just clean up the UX?

It can be.

**Comment by eyakubovich** [comment]
@xiang90 Maybe support current (but deprecate)
`etcdctl member add ID URL`

and new
`etcdctl member add --peerURLs=URL`
?

**Comment by xiang90** [comment]
@eyakubovich We need to do 

```
etcdctl member add --peerURLs=URL --ID=0xabc
```

Right? This breaks the ux, like we have `etcdctl set foo bar` not `etcdctl set --key=foo --value=bar`

**Comment by eyakubovich** [comment]
@xiang90 I actually made a typo: I meant:
`etcdctl member add --peerURLs=URL ID`.

So ID is positional, everything else is named.

**Comment by xiang90** [comment]
@eyakubovich I am fine with this approach. Any objection @jonboulle @robszumski ?

**Comment by jonboulle** [comment]
to be totally clear, ID is then the only required (positional) arg, and everything else is optional? if so, sgtm

**Comment by xiang90** [comment]
> to be totally clear, ID is then the only required (positional) arg, and everything else is optional

For delete, peerURLs is optional. For add, update, it is not optional.

**Comment by jonboulle** [comment]
Ehh. I'm not particularly fond of "required options". But I understand we're in an awkward position. I'll defer to the group.

**Comment by robszumski** [comment]
@xiang90 @jonboulle Having a required flag is probably fine.

My main concern is that folks will attempt to update the other two items in `member list`, like:

```
$ etcdctl member update a8266ecf031671f3 --name=mynewname
```

This isn't a supported operation, right? We probably need to add some code to handle it like so:

```
$ etcdctl member update a8266ecf031671f3 --name=mynewname
Member name is not a cluster-wide setting.
Changed it via the flags or env vars used to start the etcd process
```

```
$ etcdctl member update a8266ecf031671f3 --clientURLs=http://127.0.0.1:23792
Member's client URLs are not a cluster-wide setting.
Changed it via the flags or env vars used to start the etcd process
```

**Comment by robszumski** [comment]
> @gyuho Right... We might start to fill in the man page. But not a priority for now.

I think this man page is part of the problem and needs to be fixed. This is specifically about a usability problem with the tool, and folks have nowhere to go to figure out how to use it. Especially if we add some required flags.

**Pull request closed without merging**